machineconfiguration/v1/types_machineosconfig.go

```diff
@@ -0,0 +1,207 @@
+	// By separating the two secrets, the risk of write credentials becoming compromised is reduced.
+	// +required
+	RenderedImagePushSecret ImageSecretObjectReference `json:"renderedImagePushSecret"`
+	// renderedImagePushSpec describes the location of the final image.
+	// The MachineOSConfig object will use the in cluster image registry configuration.
+	// If you wish to use a mirror or any other settings specific to registries.conf, please specify those in the cluster wide registries.conf via the cluster image.config, ImageContentSourcePolicies, ImageDigestMirrorSet, or ImageTagMirrorSet objects.
+	// The format of the image push spec is: host[:port][/namespace]/name:<tag> or svc_name.namespace.svc[:port]/repository/name:<tag>.
+	// The length of the push spec must be between 1 to 447 characters.
+	// +required
+	RenderedImagePushSpec ImageTagFormat `json:"renderedImagePushSpec"`
+	// containerFile describes the custom data the user has specified to build into the image.
+	// This is also commonly called a Dockerfile and you can treat it as such. The content is the content of your Dockerfile.
+	// See https://github.com/containers/common/blob/main/docs/Containerfile.5.md for the spec reference.
+	// This is a list indexed by architecture name (e.g. AMD64), and allows specifying one containerFile per arch, up to 4.
+	// +patchMergeKey=containerfileArch
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=containerfileArch
+	// +kubebuilder:validation:MinItems=0
+	// +kubebuilder:validation:MaxItems=4
+	// +optional
+	Containerfile []MachineOSContainerfile `json:"containerFile" patchStrategy:"merge" patchMergeKey:"containerfileArch"`
+}
+
+// MachineOSConfigStatus describes the status this config object and relates it to the builds associated with this MachineOSConfig
+type MachineOSConfigStatus struct {
+	// observedGeneration represents the generation of the MachineOSConfig object observed by the Machine Config Operator's build controller.
+	// +kubebuilder:validation:XValidation:rule="self >= oldSelf", message="observedGeneration must not move backwards"
+	// +kubebuilder:validation:Minimum=0
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+	// currentImagePullSpec is the fully qualified image pull spec used by the MCO to pull down the new OSImage. This includes the sha256 image digest.
+	// This is generated when the Machine Config Operator's build controller successfully completes the build.
+	// The format of the image pull spec is: host[:port][/namespace]/name@sha256:<digest>,
+	// where the digest must be 64 characters long, and consist only of lowercase hexadecimal characters, a-f and 0-9.
+	// The length of the whole spec must be between 1 to 447 characters.
+	// +optional
+	CurrentImagePullSpec ImageDigestFormat `json:"currentImagePullSpec,omitempty"`
+	// machineOSBuild is a reference to the MachineOSBuild object for this MachineOSConfig, which contains the status for the image build.
```

Comment:
Can this change over time? If so, what happens to the old ones?

Reply:
The mapping should only be 1-to-1, and recreated (rebuilt) if the build object is deleted. cc @cheesesashimi

Reply:
Correct. If the most current MachineOSBuild is deleted, it will be recreated, starting a new build.

Reply:
So, is deleting the build something the controller might do when it wants to trigger a new build? It doesn't leave behind old builds for historical reasons?

Reply:
Yes, we have an annotation that instructs the controller to delete the current MOSB and kick off a rebuild. The idea being that e.g. a failed build should have been parsed by the user before the kick that off.

Reply:
An annotation on the MOSB? That probably could have been a spec field (we have prior art)

But ok, and, I don't remember now whether this was in the EP or not, but, is there any particular reason why you don't keep history of builds around? If I were a user and wanted to compare several historic builds, with the current behaviour, I could not do that right?

Reply:
> An annotation on the MOSB? That probably could have been a spec field (we have prior art)

We only set it as an annotation because we didn't have a field for it yet. This PR adds a status field (the `MachineOSBuild` object reference field) so we don't need to set this annotation anymore.

> is there any particular reason why you don't keep history of builds around? If I were a user and wanted to compare several historic builds, with the current behaviour, I could not do that right?

It was an oversight. I would like to eventually keep track of history but I don't know what the best practices are around that.

Reply:
> It was an oversight. I would like to eventually keep track of history but I don't know what the best practices are around that.

If that's on the cards, then having a single reference here is probably not what we want. We probably want a slice instead

```
// machineOSBuilds is a reference to the MachineOSBuild objects for this MachineOSConfig.
// This list will contain up to the last 10 builds, with the most recent builds at the top of the list.
MachineOSBuilds []MachineOSBuildHistory `json:"machineOSBuilds,omitempty"`
```

Where the historic item probably needs a little more than the name, roughly perhaps something akin to
```
machineOSBuilds:
- machineOSBuild:
    name: <mosc-name>-<suffix>
  created: <timestamp>
  finished: <timestamp>
  finalImagePushSpec: ...
  buildSummary: Completed | Failed | Pending
```

If we tie ourselves now to a single instance of an object, we can still add this slice later, but you will still always need to populate the current API

Reply:
That's perfectly acceptable as an approach, however, where we've done this before, we have had a list from the beginning, and included the latest build inside the history, see the CVO for example

You have until you ship the GA to decide if you want to change this. In theory, you could tombstone the `machineOSBuild` field between now and 4.19 shipping if you decided to fix the history thing sooner

But once the field is shipped, you must always populate it no matter what additional history you add

Reply:
Had some discussions around this. I think we're leaning to keep it as two fields:
1. machineOSBuild, which is a reference to the current active build, and
2. machineOSBuildHistory, which is a list of history objects for all builds that have completed.

The idea being that the first field is used to reference the latest active build, and the history is only updated whenever a build completes. For this iteration, we'll just have the former reference for now, and we'll add the history field at a later time once we're ready to implement that.